#include "httpinfo.h"
#include "ui_httpinfo.h"
#include <QTextCodec>
#include <QDebug>

HttpInfo::HttpInfo(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::HttpInfo)
{
    ui->setupUi(this);

    managerAddress = NULL;
    managerWeather = NULL;
    managerTime = new QNetworkAccessManager();

    connect(managerTime, SIGNAL(finished(QNetworkReply*)),
            this, SLOT(get_time(QNetworkReply*)));
}

HttpInfo::~HttpInfo()
{
    delete ui;
    delete managerTime;
    delete managerAddress;
    delete managerWeather;
}

//获取当前北京时间
void HttpInfo::on_getTimeBt_clicked()
{
    //创建请求
    QNetworkRequest request(QUrl("http://gb.weather.gov.hk/cgi-bin/hko/localtime.pl"));

    //通过管理器试图向网络获取数据，数据就绪时，managerTime将会触发readyRead()信号
    QNetworkReply *reply = managerTime->get(request);

    connect(reply, SIGNAL(readyRead()),
            this, SLOT(readReply()));
}

//获取当前天气
void HttpInfo::on_getWeatherBt_clicked()
{
    if(managerWeather == NULL)
    {
        managerWeather = new QNetworkAccessManager();
        connect(managerWeather, SIGNAL(finished(QNetworkReply*)), this, SLOT(get_weather(QNetworkReply*)));
    }

    //url地址
    QUrl url("http://www.weather.com.cn/data/cityinfo/101010100.html");
    //创建请求
    QNetworkRequest request(url);
    //通过管理器获取数据
    managerWeather->get(request);

}

//获取当前城市位置
void HttpInfo::on_getAddressBt_clicked()
{
    if(managerAddress == NULL)
    {
        managerAddress = new QNetworkAccessManager();
        connect(managerAddress, SIGNAL(finished(QNetworkReply*)), this, SLOT(get_address(QNetworkReply*)));
    }
    //url地址
    QUrl url("http://w.1616.net/chaxun/iptolocal.php");
    //创建请求
    QNetworkRequest request(url);
    //通过管理器获取数据
    managerAddress->get(request);

}

void HttpInfo::get_time(QNetworkReply* reply)
{
    QString msg = reply->readAll();
    QStringList list = msg.split("\n");
    //qDebug()<<list;
    ui->dataLebel->setText(list.at(0)+"-"+list.at(1)+"-"+list.at(2));
    ui->timeLabel->setText(list.at(3)+":"+list.at(4)+":"+list.at(5));
}
void HttpInfo::get_weather(QNetworkReply* reply)
{
    QString msg = reply->readAll();
    QJsonParseError error;
    QJsonDocument document = QJsonDocument::fromJson(msg.toUtf8(), &error);//创建一个JSonDocument对象
    if(error.error == QJsonParseError::NoError)
    {
        //没有错误说明是json数据
        QJsonObject obj =  document.object();
        QJsonValue val = obj.value("weatherinfo");
        obj = val.toObject();
        ui->textEdit->append(obj.value("temp1").toString());
        ui->textEdit->append(obj.value("temp2").toString());
        ui->textEdit->append(obj.value("weather").toString());
    }

}
void HttpInfo::get_address(QNetworkReply* reply)
{
    QString msg = reply->readAll();
    int pos = msg.indexOf("]");
    msg.remove(pos, msg.length()-pos);

    pos = msg.indexOf("[");
    msg.remove(0, pos+1);

    QStringList list = msg.split(",");

    msg = list.at(1);
    ui->addressLabel->setText(msg.remove('\''));
}

void HttpInfo::readReply()
{
    qDebug() << __FUNCTION__;
}
